lib/private/AppFramework/Middleware/NotModifiedMiddleware.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+declare(strict_types=1);
+/**
+ * @copyright Copyright (c) 2020, Roeland Jago Douma <roeland@famdouma.nl>
+ *
+ * @author Roeland Jago Douma <roeland@famdouma.nl>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OC\AppFramework\Middleware;
+
+use OCP\AppFramework\Http;
+use OCP\AppFramework\Http\Response;
+use OCP\AppFramework\Middleware;
+use OCP\IRequest;
+
+class NotModifiedMiddleware extends Middleware {
+	/** @var IRequest */
+	private $request;
+
+	public function __construct(IRequest $request) {
+		$this->request = $request;
+	}
+
+	public function afterController($controller, $methodName, Response $response) {
+		$etagHeader = $this->request->getHeader('IF_NONE_MATCH');
+		if ($etagHeader !== '' && $response->getETag() !== null && trim($etagHeader) === '"' . $response->getETag() . '"') {
```

Comment:
```suggestion
		if ($etagHeader !== '' && $response->getETag() !== null && trim($etagHeader, ' "') ===  $response->getETag()) {
```

I find that a bit easier to read but just a detail ;) Was hard to write to it should be hard to read :rofl:

Reply:
Nope. Because it is not the same. You'll have to test first if it is surrounded by "
Since the " are part of the etag.

Reply:
I know. Actually (to make it 100% correct) we should not care about the "" at this stage at all. As you said it's a part of the etag value and the code using setETag should provide a proper value. setETag should validate the input and throw a exception if the value is not quoted.

Reply:
Well that would break the rest of the API. For now lets do this. It is not the prettiest but it is tested and relatively small.